crates/state/src/state_op.rs

```diff
@@ -162,39 +172,61 @@ impl StateCache {
         let state = self.state_mut();
         assert!(slot > state.cur_slot, "stateop: decreasing slot");
         state.cur_slot = slot;
+
+        // Add diff
+        self.diff_container.cur_slot_diff = RegisterDiff::Replace(slot);
     }
 
     /// Sets the last block commitment.
     pub fn set_prev_block(&mut self, block: L2BlockCommitment) {
         let state = self.state_mut();
         state.prev_block = block;
+
+        // Add diff
+        self.diff_container.prev_block_diff = RegisterDiff::Replace(block);
     }
 
     /// Sets the current epoch index.
     pub fn set_cur_epoch(&mut self, epoch: u64) {
         self.state_mut().cur_epoch = epoch;
+
+        // Add diff
+        self.diff_container.cur_epoch_diff = RegisterDiff::Replace(epoch);
     }
 
     /// Sets the previous epoch.
     pub fn set_prev_epoch(&mut self, epoch: EpochCommitment) {
         self.state_mut().prev_epoch = epoch;
+
+        // Add diff
+        self.diff_container.prev_epoch_diff = RegisterDiff::Replace(epoch);
     }
 
     /// Sets the previous epoch.
     pub fn set_finalized_epoch(&mut self, epoch: EpochCommitment) {
         self.state_mut().finalized_epoch = epoch;
+
+        // Add diff
+        self.diff_container.finalized_epoch_diff = RegisterDiff::Replace(epoch);
     }
 
     /// Updates the safe L1 block.
     pub fn update_safe_block(&mut self, height: u64, record: L1HeaderRecord) {
         let state = self.state_mut();
         state.l1_state.safe_block_height = height;
-        state.l1_state.safe_block_header = record;
+        state.l1_state.safe_block_header = record.clone();
+
+        // Add diff
+        self.diff_container.l1_state_diff.safe_block_height_diff = RegisterDiff::Replace(height);
```

Comment:
`l1_state_diff` and similar containers need to have better interface for updates like this.